### ISSUE TITLE
TwiceDifferentiable and FMinbox seem incompatible

### ISSUE THREAD

**Issue by cstjean** [JuliaNLSolvers/Optim.jl#600]
```julia
using Optim
f(x) = (1.0 - x[1])^2 + 100.0 * (x[2] - x[1]^2)^2

lower = [1.25, -2.1]
upper = [Inf, Inf]
initial_x = [2.0, 2.0]
td = TwiceDifferentiable(f, initial_x; autodiff=:forward)
od = OnceDifferentiable(f, initial_x; autodiff=:forward)

# Both work
results = optimize(od, initial_x, lower, upper, Fminbox{GradientDescent}())
results = optimize(td, initial_x, GradientDescent())
```

This code, loosely taken from the docs, works fine on 0.14.1. However, combining `td` and  `FminBox` doesn't:

```julia
> results = optimize(td, initial_x, lower, upper, Fminbox{GradientDescent}())
MethodError: objects of type NLSolversBase.TwiceDifferentiable{Float64,Array{Float64,1},Array{Float64,2},Array{Float64,1}} are not callable

Stacktrace:
 [1] finite_difference_gradient!(::Array{Float64,1}, ::NLSolversBase.TwiceDifferentiable{Float64,Array{Float64,1},Array{Float64,2},Array{Float64,1}}, ::Array{Float64,1}, ::DiffEqDiffTools.GradientCache{Void,Void,Void,Val{:central},Float64,Val{true}}) at /home/cst-jean/.julia/v0.6/DiffEqDiffTools/src/gradients.jl:282
 [2] (::Optim.#g!#8{NLSolversBase.TwiceDifferentiable{Float64,Array{Float64,1},Array{Float64,2},Array{Float64,1}}})(::Array{Float64,1}, ::Array{Float64,1}) at /home/cst-jean/.julia/v0.6/Optim/src/objective_types.jl:8
 [3] #optimize#118(::Float64, ::Float64, ::Float64, ::Bool, ::Int64, ::Bool, ::Bool, ::Bool, ::Void, ::Int64, ::Void, ::Void, ::Float64, ::Float64, ::Float64, ::Optim.##119#128, ::Optim.Options{Float64,Void}, ::Array{Any,1}, ::Optim.#optimize, ::NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1},Val{false}}, ::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1}, ::Optim.Fminbox{Optim.GradientDescent}) at /home/cst-jean/.julia/v0.6/Optim/src/multivariate/solvers/constrained/fminbox.jl:181
 [4] optimize(::NLSolversBase.OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1},Val{false}}, ::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1}, ::Optim.Fminbox{Optim.GradientDescent}) at /home/cst-jean/.julia/v0.6/Optim/src/multivariate/solvers/constrained/fminbox.jl:145
 [5] #optimize#116(::Array{Any,1}, ::Function, ::NLSolversBase.TwiceDifferentiable{Float64,Array{Float64,1},Array{Float64,2},Array{Float64,1}}, ::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1}, ::Optim.Fminbox{Optim.GradientDescent}) at /home/cst-jean/.julia/v0.6/Optim/src/multivariate/solvers/constrained/fminbox.jl:106
 [6] optimize(::NLSolversBase.TwiceDifferentiable{Float64,Array{Float64,1},Array{Float64,2},Array{Float64,1}}, ::Array{Float64,1}, ::Array{Float64,1}, ::Array{Float64,1}, ::Optim.Fminbox{Optim.GradientDescent}) at /home/cst-jean/.julia/v0.6/Optim/src/multivariate/solvers/constrained/fminbox.jl:106
 [7] include_string(::String, ::String) at ./loading.jl:522
```

**Comment by anriseth** [comment]
Yes, `Fminbox` was written with the intention of only working with first-order methods that don't require Hessians. 
Is there a reason why you want to use `TwiceDifferentiable`s with `Fminbox`?

We could dispatch on TwiceDifferentiable and just not use the Hessian, what do you think @pkofod ?

@cstjean note that the interface of `Fminbox` has been slightly changed on `master`, so there will be some breaking changes in the next version of Optim.

**Comment by anriseth** [comment]
> I'd like to use Newton's method to tune the diagonal of a diagonal covariance matrix (i.e. I'm doing maximum likelihood). Those elements have to be positive.

Fminbox does not support Newton, unfortunately. I believe the reason is that `Fminbox` has not implemented the second-order derivatives that arise from the barrier terms. It may be an easy fix, so if you wish you can try to do so yourself and submit a PR.

If you really wish to use Newton, may I propose [ConstrainedOptim.jl](https://github.com/JuliaNLSolvers/ConstrainedOptim.jl)?
It is an interior point Newton method that we intend to add to Optim soon. Note that we don't have any autodiff functionality for the constraints yet, but if they are only box constraints it is easy to write down manually (see the examples in the README)

**Comment by cstjean** [comment]
Thank you for the suggestions, will have a look. I already solved my problem by a variable substitution: `sigma2 = exp(logsigma2)`, and I optimize `logsigma2`

**Comment by pkofod** [comment]
Having had a little look at how we could improve the code behind Fminbox, I did realize one reason why you might not want to use Newton with Fminbox.

Fminbox is basically a barrier method that uses the log-barrier function to keep iterates from moving out of bounds (primal interior point method if you will). For large barrier parameters, Newton's method might be fine, if B(x, mu) = f(x) - mu*sum(log(c_i(x)) has it's (unconstrained) minimum in the interior of the box. However, as mu decreases, x_k is allowed to approach the boundaries of the box. This is good. What is not good is that the Newton steps will tend to be outside of the box. This means that our line search will find relatively small steps, thus ruining the quadratic convergence properties (where the last steps are characterized by being full or close to full Newton steps).

The point is, that using Newton's method in our Fminbox setup could work, and we can allow for it, but it may not be a very good idea, as you're doing the expensive part of Newton, but you're not getting the quadratic convergence. The problem may be less pronounced if your box constraints are there "just to be safe", but you often expect the minima to be interior.

**Comment by pkofod** [comment]
Yes

**Comment by pkofod** [comment]
> An alternative is to make a Fminbox-type interface for box-constrained IPNewton?

I forgot this, this is an excellent suggestion!

So to follow up on my own comments above, I have since read some more about these methods, and it is indeed as I said above. Because there is very very non-linear behavior near the boundary, the Newton steps can lose a lot of their attractiveness. That said, the quasi-newton versions suffer some of the same issues, so maybe we should just add Newton. The Hessian already exists in the preconditioner we use. But it's probably better to make a simple interface to use for IPNewton because the primal-dual approach should have much better behavior at the boundary.

**Comment by pkofod** [comment]
Was fixed and documented in #858
